pbc/bragg_peak.py

```diff
@@ -1,36 +1,104 @@
 import logging
 
+import numpy as np
+import matplotlib.pyplot as plt
+from scipy import interpolate
+
+logging.basicConfig(level=0)
 logger = logging.getLogger(__name__)
 
 
-class BraggPeak:
-    def __init__(self, bragg_peak, primary_position, shifted_position, height):
-        # todo: merge those?
-        self.peak = bragg_peak
-        self.primary_pos = primary_position
-        self.shifted_pos = shifted_position
-        self.height = height
-        self.bp_data = {"bp": self.peak,
-                        "properties":
-                            {"primary_position": self.primary_pos,
-                             "shifted_position": self.shifted_pos,
-                             "height": self.height
-                             }
-                        }
-        logger.debug("Creating BraggPeak - primary position %.2f, shifted position %.2f height %.2f\nfrom\n%s" % (
-            primary_position, shifted_position, height, bragg_peak))
+class BraggPeak(object):
+    def __init__(self, bp_domain, bp_vals):
+        if len(bp_domain) != len(bp_vals):
+            raise ValueError("Domain and values have different lengths!")
+        self.spline = interpolate.InterpolatedUnivariateSpline(bp_domain, bp_vals, ext=3)
+        self.initial_position = bp_domain[np.array(bp_vals).argmax()]
+        self.current_position = self.initial_position
+        self._weight = 1.0
+        logger.debug("Creating BraggPeak...\nPrimary max position: %f\n"
+                     "Calculated spline:\n%s"
+                     % (self.initial_position, self.spline))
 
     # todo: add more class methods like __len__, __setitem__
     # https://docs.python.org/3/reference/datamodel.html
+    def __str__(self):
+        return str("{0} with position: {1} and weight: {2}".format(
+                   self.__class__.__name__, self.position, self.weight))
+
     def __repr__(self):
-        return repr(self.bp_data)
+        return repr(self.spline)
+
+    def __getitem__(self, point):
+        """Returns value for given x axis point"""
+        return self.spline(point)
+
+    @property
+    def position(self):
+        return self.current_position
+
+    @position.setter
+    def position(self, new_position):
+        self.current_position = new_position
+
+    @property
+    def weight(self):
+        return self._weight
+
+    @weight.setter
+    def weight(self, new_weight):
+        if new_weight < 0.0 or new_weight > 1.0:
+            raise ValueError("Weight should be from 0.0 to 1.0")
+        else:
+            self._weight = new_weight
+
+    def evaluate(self, x_arr):
+        """Evaluate for given domain"""
+        return self._weight * self.spline(x_arr + self.initial_position - self.current_position)
 
-    def __getitem__(self, item):
-        return self.bp_data[item]
+    def get_spread_idx(self, x_arr, val):
```

Comment:
the name of function is completely misleading. what does it have in common with spread ?